### DISCUSSION TITLE
Setting molekules and adding fragments to ketcher

### DISCUSSION THREAD

**Discussion by sasha-tsepilova** [epam/ketcher#722] [General]
Hi there!
We want to use Ketcher for visualizing molecules in our platform (you can check our GitHub [here](https://github.com/datagrok-ai/public/issues)). Now I'm trying to set molecule with the code provided in the developers manual. I made everything as described and tried standalone and remote Ketcher, but it still doesn't work. I keep getting _"Cannot read property 'setMolecule' of undefined"_ error. Would you mind helping me with some code examples or pointing out my mistakes?
![image](https://user-images.githubusercontent.com/75687119/129683153-55f7f6c6-1f17-4f5b-b87c-60aa488a1992.png)

**Comment by AndreiMazol** [comment]
Hi @sasha-tsepilova ,

I see that the error is happened  because of the fact that ketcher object is not available from your script. I guess Ketcher app and your App are hosted in different domain (because of security reasons, you  cannot get access to global objects that are defined inside Ketcher iframe in such case). 

To fix the issue please choose one of the following:
Option 1:  use the same domain for ketcher app  and your app (for example you  can host Ketcher as a part of your app)
Option 2: use React component library instead of Ready-to-run app. Please look at https://www.npmjs.com/package/ketcher-react
Option 3: use [PostMessage API](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) to communicate with IFrame (this option does not provided OOTB in Ketcher app yet, so  it is required to  be implemented by your own)

Best regards,
Andrei

**Reply by sasha-tsepilova** [reply]
Hi @AndreiMazol, 

Now I'm trying to get things working via React but still get errors. Would you mind providing some code examples?

Best wishes,
Sasha

**Reply by AndreiMazol** [reply]
Hi @sasha-tsepilova ,

could you share the list of errors? 

you can find an example of usage inside example folder: https://github.com/epam/ketcher/tree/master/example

Best regards,
Andrei

**Reply by sasha-tsepilova** [reply]
@AndreiMazol, the error isn't changing - cannot read property setMolecule of undefined. 
Given example works, but it still doesn't explain how to set molecules.

**Reply by AndreiMazol** [reply]
As far as I understand, you works with Ready-to-run application (*.zip package) instead of https://www.npmjs.com/package/ketcher-react package, doesn't it?

**Reply by sasha-tsepilova** [reply]
Firstly I did it, but now I'm using npm package and react.

**Reply by AndreiMazol** [reply]
Do you try to load molecule when Ketcher is initializing or by pressing some buttons? Perhaps, your case has been covered here: https://github.com/epam/ketcher/discussions/586

**Reply by AndreiMazol** [reply]
Hi @sasha-tsepilova ,

Please inform if #586 workaround is suitable for you. If you still have any problems with Ketcher initialization I am ready to help you.

Best regards,
Andrei

**Reply by sasha-tsepilova** [reply]
Hi, @AndreiMazol !
Yes, it helped, but I understood that I need to use Indigo, while previously, I thought Ketcher could provide this functionality without it. Thank you so much for being so helpful!

Best wishes,
Sasha
